rosidl_parser/rosidl_parser/definition.py

```diff
@@ -742,31 +803,38 @@ class IdlLocator:
 
     __slots__ = ('basepath', 'relative_path')
 
-    def __init__(self, basepath, relative_path):
+    def __init__(self, basepath: str, relative_path: str) -> None:
         """
         Create an IdlLocator.
 
-        :param str basepath: the basepath of file
-        :param str relative_path: the path relative to the basepath of the file
+        :param basepath: the basepath of file
+        :param relative_path: the path relative to the basepath of the file
         """
         super().__init__()
         self.basepath = pathlib.Path(basepath)
         self.relative_path = pathlib.Path(relative_path)
 
-    def get_absolute_path(self):
+    def get_absolute_path(self) -> pathlib.Path:
         return self.basepath / self.relative_path
 
 
+IdlContentElement: 'TypeAlias' = Union[Include, Message, Service, Action]
+IdlContentElementT = TypeVar('IdlContentElementT', bound=IdlContentElement)
+
+
 class IdlContent:
     """The content of an IDL file consisting of a list of elements."""
 
     __slots__ = ('elements', )
 
-    def __init__(self):
+    def __init__(self) -> None:
         super().__init__()
-        self.elements = []
+        self.elements: List[IdlContentElement] = []
 
-    def get_elements_of_type(self, type_):
+    def get_elements_of_type(
+            self,
+            type_: Type[IdlContentElementT]
+    ) -> List[IdlContentElementT]:
```

Comment:
I could be wrong, but shouldn't this be:

```suggestion
    ) -> List[IdlContentElement]:
```

Reply:
Due to the `isinstance` check we know that they have to match `type_` so using IdlContentElementT is correct.

```python
return [e for e in self.elements if isinstance(e, type_)]
```

This make it so when you use `get_elements_of_type` it knows that returned List only has objects of type `Action`

```python
foo = IdlContent().get_elements_of_type(Action)
reveal_type(foo)  # List[Action]
```

Reply:
I guess I don't understand then.  This method is essentially returning a subset of `self.elements`, which is typed as `List[IdlContentElement]`.  Thus it seems to me that the type of `self.elements` and the return type of `get_elements_of_type` should be the same, no?

Reply:
Ok so in line 821 is a Union of all possible entries into the `elements` list. I got these from the `extract_content_from_ast` in ros_idl.parser. These are all the possible values that can go into the `elements` list.

Reply:
Line 822 is a TypeVar that is upper bound by the TypeAlias from 821.  Because of the `isinstance` check inside the list interpolation we know that the returned List will be a subset of `self.elements` which all match `type_`. Since we know the the list in a subset of only `type_` objects we can make the function generic over `type_` and have the list match the type of `type_`. I'm sorry if my explanations are not super clear, for more information check out mypy's [page](https://mypy.readthedocs.io/en/stable/generics.html#generic-functions) about generic functions.